src/node_i18n.cc

```diff
@@ -62,6 +72,296 @@ bool flag_icu_data_dir = false;
+  } else {
+    ThrowICUError(env, status, "Unable to transcode buffer");
+  }
+}
+
+// Converts to UCS2 from ISO-8859-1 and US-ASCII
+// args[0] is the encoding to convert from
+// args[1] must be a buffer instance
+void ConvertToUcs2(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  CHECK(args[0]->IsString());
+  THROW_AND_RETURN_UNLESS_BUFFER(env, args[1]);
+
+  Utf8Value name(env->isolate(), args[0]);
+
+  SPREAD_ARG(args[1], ts_obj);
+
+  UErrorCode status = U_ZERO_ERROR;
+  size_t len;
+
+  MaybeStackBuffer<UChar> buf(ts_obj_length);
+
+  Converter from(env, *name);
+  if (from.conv == nullptr)
+    return;
+
+  len = ucnv_toUChars(from.conv,
+                      *buf, ts_obj_length << 1,
+                      ts_obj_data, ts_obj_length,
+                      &status);
+
+  if (U_FAILURE(status)) {
+    ThrowICUError(env, status, "Unable to transcode buffer");
+    return;
+  }
+  args.GetReturnValue().Set(
+    AsBuffer(env->isolate(), &buf, ts_obj_length << 1).ToLocalChecked());
+}
+
+// Convert to a named encoding from UCS2
+// args[0] is the name of the encoding being converted to
+// args[1] must be a buffer instance
+void ConvertFromUcs2(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  CHECK(args[0]->IsString());
+  THROW_AND_RETURN_UNLESS_BUFFER(env, args[1]);
+
+  Utf8Value name(env->isolate(), args[0]);
+  SPREAD_ARG(args[1], ts_obj);
+
+  UErrorCode status = U_ZERO_ERROR;
+
+  UChar* source = nullptr;
+  MaybeStackBuffer<UChar> swapspace;
+  const size_t length = ts_obj_length >> 1;
+
+  if (IsLittleEndian()) {
+    source = reinterpret_cast<UChar*>(ts_obj_data);
+  } else {
+    swapspace.AllocateSufficientStorage(length);
+    for (size_t n = 0, i = 0; i < length; n += 2, i += 1) {
+      const uint8_t hi = static_cast<uint8_t>(ts_obj_data[n + 0]);
+      const uint8_t lo = static_cast<uint8_t>(ts_obj_data[n + 1]);
+      swapspace[i] = (hi << 8) | lo;
+    }
+    source = *swapspace;
+  }
+
+  uint32_t len;
+  MaybeStackBuffer<char> buf(length);
+
+  Converter to(env, *name);
+  if (to.conv == nullptr)
+    return;
+  ucnv_setSubstChars(to.conv, "?", 1, &status);
+  status = U_ZERO_ERROR;
+
+  len = ucnv_fromUChars(to.conv, *buf, length, source, length, &status);
+
+  if (U_SUCCESS(status)) {
+    args.GetReturnValue().Set(
+      AsBuffer(env->isolate(), &buf, len).ToLocalChecked());
+  } else {
+    ThrowICUError(env, status, "Unable to transcode buffer");
+  }
+}
+
+// Converts from UTF-8 to UCS2
+// args[0] must be a buffer instance
+void Ucs2FromUtf8(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  THROW_AND_RETURN_UNLESS_BUFFER(env, args[0]);
+  SPREAD_ARG(args[0], ts_obj);
+
+  UErrorCode status = U_ZERO_ERROR;
+
+  MaybeStackBuffer<UChar> buf;
+  int32_t len;
+
+  u_strFromUTF8(*buf, 1024, &len, ts_obj_data, ts_obj_length, &status);
```

Comment:
Passing the kBufferSize in to MaybeStackBuffer is not good here because if it fails we allocate sufficient storage and try again.